### DISCUSSION TITLE
Load MD frames

### DISCUSSION THREAD

**Discussion by NAEV95** [BradyAJohnston/MolecularNodes#362] [Q&A]
Hi! I really like this project and just discovered it yesterday. I was wondering whether it would be possible to have all the frames and edit each frame when loading an MD. E.g I would like to have all the states superimposed to show the overall movement as a picture. Is there a way to do this?

**Comment by BradyAJohnston** [comment]
This feature could certainly be added. As it is, it would be a little tricky to pull it off with the nodes that are available, but when Blender 4.0 comes out I should be able to add this as a feature without too much trouble.

**Comment by NAEV95** [comment]
I made a script for this if it can be of help:
```python
import bpy

def duplicate_for_each_frame():
    # Specify the object name of your MD system
    object_name = "Name_of_Your_MD_System"
    
    obj = bpy.data.objects.get(object_name)
    if not obj:
        print(f"Object {object_name} not found!")
        return
    
    # Define the range of frames, starting from the second frame and ending at 250
    start_frame = 2  # Start from the second frame
    end_frame = 250  # Set end frame to 250

    # List to store the duplicated objects
    duplicated_objects = []

    # Modify the range to select every other frame
    for frame in range(start_frame, end_frame + 1, 2):  # Step of 2 for every other frame
        # Set the current frame
        bpy.context.scene.frame_set(frame)
        
        # Duplicate the MD object
        bpy.ops.object.select_all(action='DESELECT')
        obj.select_set(True)
        bpy.ops.object.duplicate()
        
        # Get the duplicated object and set it as the active object
        duplicated_obj = bpy.context.selected_objects[0]
        
        # Rename the duplicated object based on the frame number
        duplicated_obj.name = f"{object_name}_Frame_{frame}"
        
        # Apply all modifiers on the duplicated object
        for modifier in duplicated_obj.modifiers:
            bpy.ops.object.modifier_apply({"object": duplicated_obj}, modifier=modifier.name)

        # Add to the list of duplicated objects
        duplicated_objects.append(duplicated_obj)

        # Deselect the duplicated object
        duplicated_obj.select_set(False)

    # Join all the duplicated objects into a single object
    bpy.ops.object.select_all(action='DESELECT')
    for obj in duplicated_objects:
        obj.select_set(True)
    bpy.context.view_layer.objects.active = duplicated_objects[0]
    bpy.ops.object.join()

# Call the function to execute
duplicate_for_each_frame()
 

```

This is the final look I got in the end 
![image](https://github.com/BradyAJohnston/MolecularNodes/assets/63109284/d7e0e772-c588-4c6a-b62b-fbe0886c3cd0)
The issue is that I get extremely large file doing this and the rendering becomes quite long! (5GB)

**Comment by BradyAJohnston** [comment]
Duplicating objects will be the reason you are getting such large file sizes. This will be able to be done through geometry nodes through a more memory efficient manner, but will likely need to wait for Blender 4.0

**Comment by BradyAJohnston** [comment]
This can now be achieved in 4.0 using the 'Repat Zone', but only after importing the trajectory with `In Memory` ticked so that it all imports. This will loop through and for each iteration of the loop, get a new frame of the trajectory and then create a style, and join them all together as instances at the end.

It is quite performant, but beware with more frames realized as you increase the number of iterations it will get computationally more intense.

![CleanShot 2023-11-30 at 10 18 47@2x](https://github.com/BradyAJohnston/MolecularNodes/assets/36021261/cf5c52c2-e539-4452-8100-bdafa722def3)

![CleanShot 2023-11-30 at 10 18 59@2x](https://github.com/BradyAJohnston/MolecularNodes/assets/36021261/e8112a5e-e2ac-4055-8164-0b3c6dcb324a)

**Reply by satary** [reply]
Hello Brady! I have got an issue, looks like after changing the way trajectories are handled now the Animate frame node can not be used, as it requires the collection with frames. Are there any workarounds, or I am doing something wrong here?

**Reply by satary** [reply]
@BradyAJohnston I am sorry to bother, can you suggest some workarounds for the current MN version?
Thanks!

**Reply by NAEV95** [reply]
@BradyAJohnston btw thanks to this I made it to the cover of the journal :) https://www.google.com/url?sa=i&url=https%3A%2F%2Fpubs.acs.org%2Fdoi%2Fabs%2F10.1021%2Facs.jcim.3c00999&psig=AOvVaw2OyNdSeDLhcMe_dThnPVM8&ust=1726915880567000&source=images&cd=vfe&opi=89978449&ved=0CBQQjRxqFwoTCMCe046t0YgDFQAAAAAdAAAAABAE

**Reply by BradyAJohnston** [reply]
Hi @satary sorry for mising this earlier! Yes the import method has change to only stream from the MDAnalysis trajectory underneath. I can potentially add the method to load it all into Geometry Nodes if you would like it, but I initially dropped it because the streaming method has a bunch more features and is much more efficient. 

To somewhat replicate the multiple structures method, you can use this approach: 
![image](https://github.com/user-attachments/assets/d257a18d-58b9-490d-8b50-1e48ed0bad1d)

The `Simulation Zone` lets us save the state of the previous frame, we convert it to an `Instance` and then join it before all of the other existing Instances. Then we also select the `n` number of Instances, dropping anything that is higher. This lets us keep the previous `n` frames of the simulation.

We can also make the updating happen only every `x` number of frames by adding this to the start: 
![image](https://github.com/user-attachments/assets/f796f52f-4239-4b91-ad60-c3db9dadb4fc)


https://github.com/user-attachments/assets/090af2e9-8f64-46f9-a0dc-975d1eda78dc

**Reply by BradyAJohnston** [reply]
@NAEV95 congrats on the cover thanks for sharing! I'm glad that I was able to help :)
